### PR TITLE
Issue #4800 - fix WebSocket LinkageError and invalid PathParam type handling

### LINKED ISSUE #4800: WebSocket DistributionTests failure on JDK14
**Jetty version**
10.0.x

**Java version**
OpenJDK 14.0.1

**Description**
Test failure for `DistributionTests.testSimpleWebAppWithWebsocket()`:
https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/jetty-10.0.x/995/tests
> 2020-04-21 15:21:16.205:INFO :oejtd.DistributionTester:ConsoleStreamer/STDERR: java.lang.LinkageError: loader constraint violation: when resolving method 'void org.eclipse.jetty.tests.webapp.websocket.bad.BadOnOpenServerEndpoint.onOpen(javax.websocket.Session, org.eclipse.jetty.tests.webapp.websocket.bad.StringSequence)' the class loader 'bootstrap' of the current class, java/lang/Object, and the class loader org.eclipse.jetty.webapp.WebAppClassLoader @235a0c16 for the method's defining class, org/eclipse/jetty/tests/webapp/websocket/bad/BadOnOpenServerEndpoint, have different Class objects for the type org/eclipse/jetty/tests/webapp/websocket/bad/StringSequence used in the signature (java.lang.Object is in module java.base of loader 'bootstrap'; org.eclipse.jetty.tests.webapp.websocket.bad.BadOnOpenServerEndpoint is in unnamed module of loader org.eclipse.jetty.webapp.WebAppClassLoader @235a0c16, parent loader org.eclipse.jetty.start.Classpath$Loader @e6ea0c6)

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Issue #4800**

Changes to the WebSocket MethodHandle lookup were needed as the implementation has changed in JDK 14 and our method of using the `MethodHandles.publicLookup()` no longer works.

The type of PathParams in the endpoint signature are now being checked at deployment, where they were previously being checked during the first upgrade request to that endpoint.

### CODE REVIEW

File: tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java [hunk 1]
```diff
@@ -440,19 +439,14 @@ public void testSimpleWebAppWithWebsocket(String arg) throws Exception
...
-                assertEquals(HttpStatus.OK_200, response.getStatus());
-                assertThat(response.getContentAsString(), containsString("Hello"));
-                assertThat(response.getContentAsString(), not(containsString("<%")));
+                assertEquals(HttpStatus.SERVICE_UNAVAILABLE_503, response.getStatus());
 
                 client.GET("http://localhost:" + port + "/test2/index.jsp");
-                assertEquals(HttpStatus.OK_200, response.getStatus());
-                assertThat(response.getContentAsString(), containsString("Hello"));
-                assertThat(response.getContentAsString(), not(containsString("<%")));
+                assertEquals(HttpStatus.SERVICE_UNAVAILABLE_503, response.getStatus());
```
Comment (olamy) [inline tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:476]: I had the idea about adding another webapp to be sure it's still deployed despite having this LinkageError

Comment (lachlan-roberts) [reply tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:476]: The MethodHandle changes in this PR fix the LinkageError, but now this particular webapp should fail at deployment time anyway as the signature is wrong.

Comment (lachlan-roberts) [reply tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:476]: Should I add in another webapp here to test that it deploys successfully even if these two fail?

Comment (olamy) [reply tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:476]: yes. I would even say 2 :). I'm not sure about deployment order of contexts. maybe one first and one last. to ensure the failure do not break anything else to be deployed.

**Comment by lachlan-roberts** [comment]
Doesn't look like this fix for the MethodHandle.Lookup is going to work after all.

**Comment by lachlan-roberts** [comment]
@sbordet the method where we use `MethodHandles.lookup()` and drop all lookupModes except PUBLIC does work fine on JDK 14, however it doesn't work for JDK 11 if done from a named module. All access is dropped when teleporting from a named module to another module except if you are using the publicLookup. So our resulting Lookup after the `Lookup.in()` has no access permissions at all.

So the  `publicLookup()` method only works before JDK14, and `MethodHandles.lookup()` + `dropLookupModes()` only works on JDK14.

**Comment by lachlan-roberts** [comment]
@sbordet this is ready for review. The WebSocket DistributionTest has been disabled on JDK 14 and now it is passing the build.

File: tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java (hunk 1 above)
Comment (olamy) [reply tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:476]: @lachlan-roberts great tests changes! we now even have real websocket tests in the distribution tests module!

**Review by sbordet — APPROVED** [review]

**Pull request merged**